### ISSUE TITLE
DOC: Rewrite manual for parallel 

### ISSUE THREAD

**Issue by amitmurthy** [JuliaLang/julia#19579]
One of the issues with the parallel manual is that it leads with low-level API calls, viz., `remotecall`, `@spawn` and friends. New users coming into explore parallelism in Julia should probably be first exposed to high-level interfaces and packages. This issue is to track and discuss a better way to introduce Julia parallelism in the documentation.

Specifically talk about the following first:

In Base:
- `pmap`
- Multi-threading
- SharedArrays

External Packages
- DistributedArrays.jl
- MPI.jl
- A note on various projects under JuliaGPU

Typical distributed computing needs ought to be met by the above modules/packages.

This should be followed by
- a section on the in-built clustering architecture
- remotecalls, `@spawn` and Futures. This is to be presented as required only for more fine-grained control on the parallelism, or for library writers.
 
Any thoughts/comments?

**Comment by quinnj** [comment]
Here are some thoughts from someone who certainly doesn't have a background or expertise, but occasionally wants/needs to utilize some of the parallel code:

* It'd be helpful to start off with just a simple "parallel" primer to the extent of explaining the difference between: 1) green-threading, 2) multi-threading, and 3) full-blown multi-core/distributed processing.
* For a lot of my cases as a "simple user", I'm interested in the green-threading functionality, when it's wise to use it, when I can expect benefits, and an example or two of how to architect things correctly; I'm thinking particularly around `Task`, `Condition`, `notify`, `schedule`, `@async`, `Channel`s, and `wait`
* As a "simple user", I'm also not as interested (often) in learning an entirely new external package, but maybe just want to use some Base functionality to ensure that my package can run multi-core. So some simple examples/techniques of using the Base functionality (Pmap, RemoteChannel, `@everywhere`, `@spawn`, `@fetch`, `SharedArray`).
* I think after this ^ brief introduction to Base functionality, it would be good to do a more-full discourse on available external packages and their benefits. This keeps the manual focused on Base with some good primers and quick examples, while leading more experienced users to the external packages to utilize the advanced functionality.

Happy to help put together some of my own examples I mentioned above of simple cases of green-threading and simple multi-core Base usage.

**Comment by oscardssmith** [comment]
As simpler user I completely agree.

**Comment by jingpengw** [comment]
As I know, multithreading is still experimental and have some performance issues. putting threading to the end of documentation might be better?

**Comment by yuyichao** [comment]
Threading is experimental but afaik there isn't any threading specific performance issue.

**Comment by ViralBShah** [comment]
The threading api will evolve, and I/O is not thread-safe at the moment. But if you use it for pure computational tasks, I agree - there are no known performance issues.

**Comment by jingpengw** [comment]
thanks for the explanations! I was out-dated.

**Comment by StefanKarpinski** [comment]
Documentation is not technically release-blocking, there's an increasing sense that much of the distributed computing infrastructure should be in a standard package so that we can iterate on it.

**Comment by schnaitterm** [comment]
Reading through the parallel docs (this might reach beyond the scope of this issue), `Threads.@threads for ...` seems redundant with `@parallel() for ...` especially because it seems like the `@parallel` construct for for loops works exactly like OMP pragmas in C++, except that here it is part of Base.Distributed and works only with the cluster model.  Shouldn't this be a bit more integrated so that `Threads` would be an option for the backend of `@parallel`? This would make parallel code more portable and scalable, potentially.

**Comment by DokFaust** [comment]
Hello everyone! I saw this as a decent opportunity to start contributing to the project. 
Anyway most of the examples mentioned in the first two posts where present in the first version of the doc, it was just necessary to refactor the various paragraphs and add more examples.
As a first into I tried to make a distinct difference between distributed processing reliant on the runtime lybrary and "real" multi-threading presented by the `Threads` module.
Also I added in the end a brief intro to some of the most useful external packages in julia distributed programming, the code extracts are all taken from [Valentin Churavy's notes](https://github.com/stevengj/18S096/blob/master/lectures/lecture5/Parallelism.ipynb).
